x-pack/plugins/enterprise_search/public/applications/shared/use_local_storage/use_local_storage.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useState, useEffect } from 'react';
+
+export const useLocalStorage = <T>(key: string, defaultValue: T) => {
+  const [item, setItem] = useState<T>(getFromStorage());
+
+  function getFromStorage() {
```

Comment:
consistency nit - function style

```suggestion
  const getFromStorage = () => {
```

Reply:
fc877ad7dcd1da114532207f238ad9f04f5fd4d3

Reply:
3e903957fe3d49ec8275f1b212a1371aae4d0343

Reply:
Sorry, I'm confused by this change - why move `const getFromStorage = ()` below `const [item, setItem]`? I thought `const` declarations got hoisted the same way `function` does - is that not the case?

Reply:
Ah sorry just tried it myself and it is indeed a hoisting issue (SUPER weird/interesting).

Can we move the `const [item, setItem]` to line 51 in that case (below all the `*To/FromStorage` methods) and directly above the return? I think it fits a bit better to organize it that way 🤔

Reply:
Yeah, organizationally, I think you're right. But at the same time, `updateFromStorage` and `saveToStorage` both use `setItem`, so it feels more natural to keep them above those methods. Thoughts?

Reply:
Haha, oh my gosh. I have a full on circular dependency situation. Trying to make the changes you suggested below, literally no way to order them and have it compile. Going to have to rework this a bit.

Reply:
Arggh sorry 🤦‍♀️ Definitely feel free to put this back as a function() if it's a headache, just maybe add a quick comment to note it has to be that way for hoisting reasons. Although I'm a little befuddled as to how tf the hoisting is working correctly in this case lol

Reply:
7d3ec09eb5a635274c79015d71ae09e009399ccf

Reply:
I reworked it a bit so that the purely deal with localStorage. All of the state code is cleanly separated below. Should be better, seems to clean things up.